### ISSUE TITLE
Global endpoint filter

### ISSUE THREAD

**Issue by WeihanLi** [dotnet/aspnetcore#43237]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Is your feature request related to a problem? Please describe the problem.

Since we introduced the endpoint filter feature, is there also a global endpoint filter available?

So that we could avoid registering the endpoint filter for each endpoint.

### Describe the solution you'd like

Maybe likes MVC global filters, and maybe for a `Group` (`MapGroup`) likes the filter for controller

**Comment by davidfowl** [comment]
The way to do this is with groups. Make a global group and stick all of the endpoints in it.

**Comment by WeihanLi** [comment]
Thanks @davidfowl

**Comment by halter73** [comment]
I love using an empty group prefix for this! Support for gobal metadata and filters, was the main reason I figured it was worth allowing empty group prefixes. I think we can do better though. What do we think about having `WebApplication` implement `IEndpointConventionBuilder`.

## Background and Motivation

See above.

## Proposed API

namespace Microsoft.AspNetCore.Builder;

```diff
namespace Microsoft.AspNetCore.Http;

- public sealed class WebApplication : IHost, IApplicationBuilder, IEndpointRouteBuilder, IAsyncDisposable
+ public sealed class WebApplication : IHost, IApplicationBuilder, IEndpointRouteBuilder, IEndpointConventionBuilder, IAsyncDisposable
```

## Usage Examples

```C#
var app = WebApplication.Create(args);

app.AddEndpointFilter((context, next) =>
{
    return next(context);
});

app.MapGet("/", () => "Hello World!");

app.Run();
```

## Alternative Designs

Leave it as is. You can always create an route group with an empty prefix if you want this functionality as @davidfowl suggests:

```C#
var app = WebApplication.Create(args);
var routes = app.MapGroup("");

routes.AddEndpointFilter((context, next) =>
{
    return next(context);
});

routes.MapGet("/", () => "Hello World!");

app.Run();
```

## Risks

It would be too noisy to have too many extension methods that all work on `WebApplication`.

**Comment by WeihanLi** [comment]
The `MatchGroup` with empty had to update the existing code, if there are a lot of endpoints it may need a big effort to update.
Personally, I prefer the proposal

**Comment by davidfowl** [comment]
I don't like this proposal I think top level intellisense will be too noisy.

**Comment by davidfowl** [comment]
Here's an alternative proposal:

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class WebApplication 
{
+    public RouteGroupBuilder Routes { get; }
}
```

We expose a default top level `RouteGroup`  that you can hang your routes on. Internally, we'd create a default route group and   would use it for both top level routes hanging off the `WebApplication` itself and for the Routes property.

```C#
app.Routes.MapGet("/", () => "Hello");
```

is the same as:

```C#
app.MapGet("/", () => "Hello");
```

Adding global conventions is then:

```C#
app.Routes.AddEndpointFilter((context, next) =>
{
    return next(context);
});

**Comment by halter73** [comment]
Another alternative is adding a `Conventions` property that's just an `IEndpointConventionBuilder`. Under the covers, we would use groups to implement this, but we don't need to expose that implementation detail if we don't want to.

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class WebApplication 
{
+    public IEndpointConventionBuilder Conventions { get; }
}
```

Usage:

```C#
app.MapGet("/", () => "Hello");

app.Conventions.AddEndpointFilter((context, next) =>
{
    return next(context);
});
```

We've said that we don't want APIs targeting `RouteGroupBuilder` directly. It's a sealed type. Any single API should target either `IEndpointRouteBuilder` or `IEndpointConventionBuilder`, but not both. If later on, we feel this is a mistake, we could create a new interface that implements both. We could then have `RouteGroupBuilder` implement that and tell people to target the new combined interface, but I don't think we're going to need that.

**Comment by halter73** [comment]
Or we could do a hybrid approach, so we still allow you to delineate between the `IApplicationBuilder` and the `IEndpointConventionBuilder` if you want to.

```diff
namespace Microsoft.AspNetCore.Http;

public sealed class WebApplication 
{
+    public IEndpointRouteBuilder Routes { get; }
+    public IEndpointConventionBuilder Conventions { get; }
}
```

Usage:

```C#
app.Routes.MapGet("/", () => "Hello");

app.Conventions.AddEndpointFilter((context, next) =>
{
    return next(context);
});
```

**Comment by davidfowl** [comment]
I considered this as well but I figured 2 birds with one stone (see the blog posts). Is there any reason why we can't expose a sealed type? Are you predicting extension methods targeting this?

**Comment by halter73** [comment]
> I considered this as well but I figured 2 birds with one stone (see the blog posts).

I just read them before I suggested the hybrid proposal that addresses the concern of both blogposts. I think it's better because it's not adding yet a new property that arguably implements too many interfaces on a single type.

> Are you predicting extension methods targeting this?

I hope not. If we were willing to make a major change to the route group APIs in the 11th hour of .NET 7 we could change `RouteGroupBuilder` itself to not implement two very distinct interfaces. Arguably, this would avoid repeating the mistake we made with `WebApplication`:

```diff
namespace Microsoft.AspNetCore.Routing;

- public sealed class RouteGroupBuilder : IEndpointRouteBuilder, IEndpointConventionBuilder
+ public sealed class RouteGroupBuilder
{
+     public IEndpointRouteBuilder Routes { get; }
+     public IEndpointConventionBuilder Conventions { get; }
}
```

With this API change, the current workaround for global convetions goes from being like:

```C#

var app = WebApplication.Create(args);
var routes = app.MapGroup("");

routes.AddEndpointFilter((context, next) =>
{
    return next(context);
});

routes.MapGet("/", () => "Hello World!");

app.Run();
```

To something like this:

```C#
var app = WebApplication.Create(args);
var group = app.MapGroup("");

group.Conventions.AddEndpointFilter((context, next) =>
{
    return next(context);
});

group.Routes.MapGet("/", () => "Hello World!");

app.Run();
```

@andrewlock and @khalidabuhakmeh do either of you have any thoughts on this?

**Comment by halter73** [comment]
API Review Notes:

- Having to type the extra `.Routes` or `.Conventions` characters is annoying.
- Can we use destructors?
- Can we use it for RouteGroupBuilder?
- Can we remove the prefix parameter from MapGroup if this is the preferred way to add conventions to a group of endpoints?

Or we could do a hybrid approach, so we still allow you to delineate between the `IApplicationBuilder` and the `IEndpointConventionBuilder` if you want to.

**Comment by andrewlock** [comment]
> do either of you have any thoughts on this?

I mean, I have definite thoughts if we could turn back time and have `WebApplication` not directly implement `IEndpointRouteBuilder` (as I mentioned in the blog post), but I assume that's not really on the cards 😄

For the hybrid approach:
I think having `WebApplication` directly implement `IEndpointRouteBuilder` _and_ having a `Routes` property _may_ be worse than one or the other, as it adds confusion/ambiguity (i.e. the difference between `app.MapGet()` and `app.Routes.MapGet()` is not obvious)

RE splitting `RouteBuilder`:
I'm not sure if it adds a lot splitting `IEndpointRouteBuilder` and `IEndpointConventionBuilder` other than make things a bit more verbose. I quite like that I didn't even appreciate that there were two interfaces involved here previously, as everything just flows quite nicely. 

For me the top-level `Conventions` property makes most sense probably, though I'm not a fan of the name (I don't think of filters etc as being "conventions", and nor do any of the docs, even if that's now they're implemented).

Personally, I think the original workaround, of creating a "top-level route" is the best option given the existing constraints

**Comment by khalidabuhakmeh** [comment]
I agree with @andrewlock in the sense that `Conventions` doesn't make a lot of sense here. I would consider it as part of the request pipeline, so possibly `Pipeline` or `Filters` (to stay in line with Razor Pages and MVC) would make the most sense.

The thing I like about ASP.NET Core is the shared model, and that includes the shared mental model across all the approaches. Trying to be too different might drive people to "pick sides", rather than realizing the strength is in flowing between Razor Pages, Minimal API endpoints, MVC, and SignalR.
